Lib/test/test_subprocess.py

```diff
@@ -1477,6 +1477,64 @@ def test_run_kwargs(self):
         self.assertEqual(cp.returncode, 33)
 
 
+    def test_run_interrupted_cleanup_succeeds(self):
+        # In the event of KeyboardInterrupt, run() should allow the
+        # child time to cleanup (if cleanup_timeout was given).
+        # In this case, child's cleanup completes before cleanup_timeout runs out.
+        self._test_run_interrupted(0.2, 1.0)
+
+    def test_run_interrupted_cleanup_timed_out(self):
+        # In the event of SIGINT, run() should allow the
+        # child time to cleanup (if cleanup_timeout was given).
+        # In this case, child's cleanup takes too long for the cleanup_timout.
+        self._test_run_interrupted(3.0, 0.1)
+
+    @unittest.skipIf(mswindows, "cannot use setpgid, killpg, etc. on windows")
+    @unittest.skipIf(threading is None, "threading required")
+    def _test_run_interrupted(self, cleanup_time, cleanup_timeout):
+        # In the event of KeyboardInterrupt, run() should allow the
+        # child time to cleanup (if cleanup_timeout was given).
+
+        # Change the current process group to avoid
+        # interrupting our parent process.
+        original_pgid = os.getpgrp()
+        os.setpgid(0, 0)
```

Comment:
Hum, I dislike playing with process groups in an unit test. I would prefer to only send SIGINT directly to processes using their pid.